### ISSUE TITLE
Has to relaunch the dev server to have classes applied with Snowpack

### ISSUE THREAD

**Issue by samueldrai** [tailwindlabs/tailwindcss-jit#126]
### What version of @tailwindcss/jit are you using?

0,1,4

### What version of Node.js are you using?

14.2.0

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### Reproduction repository

https://www.youtube.com/watch?v=G6d-SXdcJXs

Hi there ☺️

I tried to use this new JIT feature with my a new React app built with Snowpack. 
Unfortunately, I cannot have Snowpack use Tailwind JIT properly. 

I think I might have missed something about the configuration process.
When changing a class in a component, I have to stop and relaunch the dev server. 
Here are my config files: 

```javascript
// snowpack.config.js
module.exports = {
  mount: {
    public: { url: '/', static: true },
    src: { url: '/dist' },
  },
  plugins: [
    '@snowpack/plugin-react-refresh',
    '@snowpack/plugin-dotenv',
    '@snowpack/plugin-typescript',
    '@snowpack/plugin-postcss',
  ],
  routes: [{ match: 'routes', src: '.*', dest: '/index.html' }],
  optimize: {
    /* Example: Bundle your final build: */
    // "bundle": true,
  },
  packageOptions: {},
  devOptions: {},
  buildOptions: {},
}
```

```javascript
// tailwind.config.js
module.exports = {
  purge: ['./src/**/*.{js,jsx,ts,tsx}', './public/index.html'],
  darkMode: false,
  theme: {
    extend: {},
  },
  variants: {
    extend: {},
  },
  plugins: [],
}
```

```javascript
// postcss.config.js
module.exports = {
  plugins: {
    '@tailwindcss/jit': {},
    autoprefixer: {},
  },
}

```

Thanks!

**Comment by adamwathan** [comment]
Hey thanks! In this case a reproduction repository would definitely be helpful so we don't have to try and set everything up from scratch 🙏 I've never used Snowpack before so without a reproduction repository it's going to be a lot of work for me to try and recreate this problem by hand.

**Comment by samueldrai** [comment]
Of course, there it is: `https://github.com/samueldrai/tailwind-jit-snowpack`

I thought that the config files would be enough to solve the problem.

By the way, thanks a lot for the quick answer 🚀 :)

**Comment by Xeevis** [comment]
Hello. If I may chip in I can't get the JIT watcher fully working with Snowpack either. Very simple repo https://github.com/Xeevis/tailwindcss-jit-snowpack

```bash
pnpm install
pnpm watch
```

When I change the `app/app.css` it's correctly reflected in `build/app.css` right away, changing the `app/index.html` however won't change the css on it's own, I either have to run the watch again or manually save the unchanged `app/app.css`.

**Comment by Renerick** [comment]
Can confirm same behavior with `rollup -c -w` (no snowpack) - styles are only regenerated when I save CSS files, not template files. Repro - https://github.com/Renerick/svelte-template/tree/tailwindcss-jit, `npm run dev`

**Comment by apvarun** [comment]
You can explicitly configure TAILWIND_MODE to be `watch`.

Set the env variable as `export TAILWIND_MODE="watch"`
or configure in yoru npm scripts

```
// package.json
"scripts:{
  "dev": "...",
  "watch": "TAILWIND_MODE='watch' npm run dev"
}
```

This environment variable seems to solve the problem.

**Comment by Renerick** [comment]
Interesting. This does indeed resolves the problem (with rollup at least), but I wonder if it's correct behavior, considering that it still automatically generates new styles when changing CSS files

**Comment by adamwathan** [comment]
Yeah that's correct behavior, the template watching is handled by us but the CSS file watching is handled by the build tool which explains why you're seeing that. We'll figure out how to make this more intuitive/automatic over time for sure.

**Comment by samueldrai** [comment]
Hi, 
Thanks a lot for your answers. 
I tried the solution provided by @apvarun but I still run in the same issue. 
I changed my `scripts` tag for this: 

```  "scripts": {
    "start": "snowpack dev",
    "build": "snowpack build",
    "watch": "TAILWIND_MODE='watch' yarn start"
  }```

Can't figure out why I still have this issue :/

**Comment by adamwathan** [comment]
@samueldrai just cloned your repo but it looks like Tailwind isn't even listed as a dependency, nor is there a tailwind.config.js file or anything like that so I'm not sure where to start here. I feel like a moron saying "you need to install Tailwind for Tailwind to work" but that's about the only advice I can offer currently, haha... Can you update the repo with your latest efforts?

**Comment by Xeevis** [comment]
No dice here with [snowpack](https://www.snowpack.dev/), since I'm on Windows tried to change env with [cross-env](https://www.npmjs.com/package/cross-env)
```
"scripts": {
  "dev": "cross-env NODE_ENV=development TAILWIND_MODE=watch snowpack dev",
  "watch": "cross-env NODE_ENV=development TAILWIND_MODE=watch snowpack build --watch"
},
```
Doesn't work in neither dev or watch. Template save won't trigger change in the css, I still need to bump css file for html changes to take effect.

My code https://github.com/Xeevis/tailwindcss-jit-snowpack

**Comment by samueldrai** [comment]
@adamwathan sorry for the back and forth. The problem was that the repo was a repo from my company and I could not share it directly. 😅 
Anyway, here's the up-to-date repo!
https://github.com/samueldrai/tailwind-jit-snowpack

**Comment by bramaudi** [comment]
@samueldrai  I also trying the same thing in my PHP project using `@snowpack/plugin-run-script` and thats works great. Maybe you should try.

`npm i -D @snowpack/plugin-run-script postcss-cli`

``` js
// snowpack.config.js
// ...
plugins: [
    [
      '@snowpack/plugin-run-script',
      {
        cmd: 'postcss ./app/assets/css/style.css -o ./public/css/style.css --no-source-map', // production build command
        watch: 'postcss --watch ./app/assets/css/style.css -o ./public/css/style.css --no-source-map', // (optional) dev server command
      },
    ],
  ],
// ...
```

then just `npx snowpack dev`, since tailwindcss jit already build production-ready css then no build command needed.

**Comment by Xeevis** [comment]
@bramaudi this workaround does work better, thanks! 

However when I remove classes from template file, it won't get purged from the css, it only appends new classes. For purge I need to bump save CSS file. This doesn't feel right, is this expected behavior with the JIT @adamwathan ?

https://user-images.githubusercontent.com/5835044/111964595-10e51000-8af5-11eb-8cdd-3c7214287940.mp4

**Comment by adamwathan** [comment]
@Xeevis Yeah that's been discussed before, it's a deliberate performance optimization that assumes people are always doing a fresh build when actually building for production:

https://github.com/tailwindlabs/tailwindcss-jit/issues/57

It is surprising to me to learn that people run a watcher in development and are actually committing the output of the watcher, I thought everyone was building assets like this at deploy 😲

**Comment by thomasruiz** [comment]
Hey @adamwathan, I set another repo up with the changes so you can try it out: https://github.com/thomasruiz/snowflake-tailwind

If I try to change App.tsx and change the bg-blue class, the browser refreshes, but the classes are not generated.

**Comment by austinorr** [comment]
Tailwind noob here, but I'm also excited about this jit effort and having trouble with the development workflow for `tailwind` in `snowpack`. It may not be possible (or at least, it may not be easy), but I had partly expected that the dev behavior of the jit compiler would be able to be configured to 'watch' the files matching the tailwind.config purge.content. In my case, I don't actually wish to 'watch' any css files at all since my `tailwind` classes are in my markup and templates.

I tried using the script solution proposed by @bramaudi to trigger postcss to watch my js and html files, but it hangs immediately on any the non-css syntax e.g., a js file (should have mentioned I'm also a postcss noob too). Of course, I also tried it as intended, watching my projects css file.

I also tried leveraging the `snowpack build --watch` command (which is just like the snowpack dev command except that files actually get written to the build directory; I thought re-triggering builds would trickle down through the plugins and actually re-apply the tailwind purge.content directive), but alas the `tailwind` jit compiler fires only once during startup and never again. I should note here that this is the same behavior I get when using `tailwind` straight. I should also note that I can build my project with slim css files easily as a one-off, but I cannot (yet) watch/develop my project without using the xxl tailwind css.

I'm hoping there is a way to get the jit to fire again when I edit my templates or markup files, or maybe there's some other argument or incantation I missed that can get it working as well as it clearly does in webpack.

**Comment by Xeevis** [comment]
This may be a compatibility problem with the `@snowpack/plugin-postcss` and JIT watcher, sadly javascript is not my turf either. However In the [related issue](https://github.com/snowpackjs/snowpack/issues/2916) on snowpack repo there is a workaround. Basically just create a simple snowpack plugin that will watch template files for changes and mark CSS as changed *(not drag & drop, uses hardcoded values)*.

```js
const path = require('path');
const plugin = (snowpackConfig, _pluginOptions) => {
  const appCssPath = path.join(snowpackConfig.root || process.cwd(), 'src/App.css');
  return {
    name: '@local/snowpack-css-refresh-plugin',
    onChange({filePath}) {
      if (!filePath.endsWith('.html')) {
        return;
      }
      this.markChanged(appCssPath);
    },
  };
};

module.exports = plugin;
```
In package.json
```json
  "devDependencies": {
    "@local/snowpack-css-refresh-plugin": "file:./dev/snowpack-css-refresh-plugin"
  }
```

**Comment by austinorr** [comment]
@Xeevis the local plugin idea works a charm. Don't forget to `npm install` after modifying the devDeps. Now we just need to find a way to pass the desired 'watch css filepath' and file match pattern (e.g., `"./src/**/*.{js,html,jsx,css}"`  to the plugin to avoid the hard coding. Thank you!

**Comment by Xeevis** [comment]
@austinorr I've managed to use tailwind's purge globs. JIT works pretty well for me now with Snowpack, only thing left to do is to get rid of the hardcoded CSS file. I guess only way to do that is to read and determine which CSS files hold the Tailwind imports.

Anyway here is video using purely `snowpack build --watch` with `@snowpack/plugin-postcss` and custom plugin `snowpack-tailwindcss-jit` that can be found in [my repo](https://github.com/Xeevis/tailwindcss-jit-snowpack/tree/feature/snowpack-tailwindcss-jit)

https://user-images.githubusercontent.com/5835044/112697748-88fe5d80-8e88-11eb-9c87-048b50b28d8f.mp4

**Comment by Xeevis** [comment]
I've published new Snowpack plugin that should help with this issue without any extra configuration.  Can you guys test it if it works for you too? https://github.com/jadex/snowpack-plugin-tailwindcss-jit

**Comment by verbose9** [comment]
## Here's the reproduction repo
[🎉 Repo](https://github.com/verbose9/tlwind)
`npm i` to install deps.
`npm start` to start dev server

**Comment by verbose9** [comment]
Why #181 is closed?

**Comment by verbose9** [comment]
Possible reasons might be: Snowpack caches everything indefinitely. That's why it gives ⚡ fast performance from the beginning. And in my opinion the whole point of JIT is already built into Snowpack.

[Office website](http://snowpack.dev)

**Comment by adamwathan** [comment]
The problem with Snowpack is it doesn't have support for PostCSS's dependency message API. We are working on a PR to Snowpack to fix this.

**Comment by verbose9** [comment]
> The problem with Snowpack is it doesn't have support for PostCSS's dependency message API. We are working on a PR to Snowpack to fix this.
> 
> > Why #181 is closed?
> 
> Because it's a duplicate of this issue and we only want one issue per problem.

Really happy that you guys are working on it😄😄

**Comment by verbose9** [comment]
what about that PR that you are working on??

**Comment by adamwathan** [comment]
Dude if you ever send me a message in it with two question marks in a row again I am gonna ban you from our repositories :/ We are working hard, we have 500 things to work on, this is one of them. Relax.

**Comment by deleonio** [comment]
Issue similar to https://github.com/tailwindlabs/tailwindcss-jit/issues/202
